### PR TITLE
MBExpression Lookup Expressions Implementation

### PR DESCRIPTION
**Pull request by johngschulz**

### CODE REVIEW

File: modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/function/ListSizeFunction.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.List;
+
+/**
+ * Returns the size of a given list
+ */
+public class ListSizeFunction extends FunctionExpressionImpl {
+
+    public static FunctionName NAME = new FunctionNameImpl("listSize",
+            parameter("size", Integer.class),
+            parameter("list", List.class));
```
Comment (tbarsballe) [inline modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/function/ListSizeFunction.java]: You could support Collection here, rather then merely List (List is a subclass of Collection). Wouldn't make a difference for the current functionality, but could be useful in the future.

File: modules/unsupported/mbstyle/src/test/resources/org/geotools/mbstyle/expressionParseTest.json [hunk 2]
```diff
@@ -155,6 +155,54 @@
...
+            "id": "lengthExpression",
+            "type": "symbol",
+            "source": "composite",
+            "source-layer": "waterway",
+            "layout": {
+                "text-field": ["length", [1, 2, 3, 4, 5, 6, 7, 8, 9]]
+            },
+            "paint": {
+                "text-color": ["length", "123456789"]
+            }
```
Comment (tbarsballe) [inline modules/unsupported/mbstyle/src/test/resources/org/geotools/mbstyle/expressionParseTest.json]: Tests look good overall, but could do with more nested structures, e.g:

```
["at", 0, ["get", "text-color", {"text-color": ["length", "123456789"]}]]
```
(should return `"length"`, in theory)

**Comment by johngschulz** [comment]
Addressed comments and pushed changes.

File: modules/unsupported/mbstyle/src/test/java/org/geotools/mbstyle/expression/MBLookupTest.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        Object on = getExpressionEvaluation(jn, "halo-color");
+        assertEquals(true, on);
+        // Verify that a "has" lookup expression transforms correctly to SLD.
+        MBStyle getTest = MBStyle.create(mbstyle);
+        SymbolMBLayer rgbLayer = (SymbolMBLayer) getTest.layer("hasExpression");
+        List<FeatureTypeStyle> getFeatures = rgbLayer.transformInternal(getTest);
+        try {
+            String xml = new SLDTransformer().transform(getFeatures.get(0));
+            assertTrue(xml.contains("<ogc:Function name=\"PropertyExists\"><ogc:Literal>name</ogc:Literal></ogc:Function>"));
+        } catch(Exception e) { }
```
Comment (emerkle826) [inline modules/unsupported/mbstyle/src/test/java/org/geotools/mbstyle/expression/MBLookupTest.java]: Unless this test expects to throw an Exception here, I think it's better to declare the test method `throws Exception` and don't bother with the try-catch here.

File: modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/expression/MBLookup.java [hunk 4]
```diff
@@ -33,27 +36,88 @@ public MBLookup(JSONArray json) {
...
      */
-    public Expression lookupAt(){ return null;}
+    public Expression lookupAt(){
+        // requires an instance of a "literal" array expression
+        if (json.size() == 3 && json.get(2) instanceof JSONArray){
+//            Expression e = ff.literal(json.get(2));
+            Expression e = parse.string(json,2);
+            if (e.evaluate(null, JSONArray.class) != null) {
+                JSONArray arr = e.evaluate(null, JSONArray.class);
+                Integer at = parse.string(json, 1).evaluate(null, Integer.class);
```
Comment (emerkle826) [inline modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/expression/MBLookup.java]: I think all of the calls to `evaluate` in this class will need to be taken out, and delayed to render time. You likely need stick with unevaluated expressions in here, and defer them to an existing or custom function that does the evaluation at render time (similar to the comments Torben made on https://github.com/johngschulz/geotools/pull/3)

File: modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/expression/MBLookup.java [hunk 5]
```diff
@@ -33,27 +36,88 @@ public MBLookup(JSONArray json) {
...
+    public Expression lookupGet() {
+        if (json.size() == 2 || json.size() == 3) {
+            if (json.size() == 2) {
+                Expression property = parse.string(json, 1);
+                String s = property.evaluate(null, String.class);
+                return ff.property(s);
+            }
+            if (json.size() == 3) {
+                Expression property = parse.string(json, 1);
+                String s = (property).evaluate(null, String.class);
```
Comment (emerkle826) [inline modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/expression/MBLookup.java]: same here with the calls to `evaluate`

File: modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/expression/MBLookup.java [hunk 6]
```diff
@@ -33,27 +36,88 @@ public MBLookup(JSONArray json) {
...
+            if (json.size() == 2) {
+                Expression value = parse.string(json, 1);
+                String s = value.evaluate(null, String.class);
+                return ff.function("PropertyExists", ff.literal(s));
+            }
+            if (json.size() == 3) {
+                String svalue;
+                Expression value = parse.string(json, 1);
+                if (value.evaluate(value) instanceof String){
+                    svalue = value.evaluate(null, String.class);
```
Comment (emerkle826) [inline modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/expression/MBLookup.java]: here too

File: modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/expression/MBLookup.java [hunk 7]
```diff
@@ -62,7 +126,17 @@ public MBLookup(JSONArray json) {
      *   ["length", array]: number
      * @return
      */
-    public Expression lookupLength(){ return null;}
+    public Expression lookupLength() {
+        Expression e = parse.string(json,1);
+        if (e.evaluate(e) instanceof String) {
```
Comment (emerkle826) [inline modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/expression/MBLookup.java]: possibly here as well

**Comment by johngschulz** [comment]
Addressed the issue of premature evaluations by using custom functions - I left 'lookupLength' as is for basic validation before returning the functions.

File: modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/expression/MBLookup.java (hunk 7 above)
Comment (johngschulz) [reply modules/unsupported/mbstyle/src/main/java/org/geotools/mbstyle/expression/MBLookup.java]: I left this as is for basic validation for string vs array before returning the appropriate function.

**Comment by johngschulz** [comment]
Fixed up my merge conflict, and tests are all working.

**Pull request merged**